### PR TITLE
[10.0][IMP] sale_order_type: find type by rule

### PR DESCRIPTION
**Pull request by SimoRubi**
Add feature to automatically select the sale order type based on products (or their categories) of the sale order.

### CODE REVIEW

**Comment by SimoRubi** [comment]
Travis is failing due to #717

**Comment by SimoRubi** [comment]
Should be rebased after https://github.com/OCA/sale-workflow/pull/718 is merged

**Comment by SimoRubi** [comment]
Rebased after #718

File: sale_order_type/readme/USAGE.rst
Comment (eLBati) [inline sale_order_type/readme/USAGE.rst:2]: Add a point about the new `find by rule` button

File: sale_order_type/models/sale_order.py
Comment (eLBati) [inline sale_order_type/models/sale_order.py:39]: You should probably use `order_line.product_id`

File: sale_order_type/views/sale_order_view.xml
Comment (eLBati) [inline sale_order_type/views/sale_order_view.xml:13]: This should be also done for invoices

**Review by eLBati — CHANGES_REQUESTED** [review]

**Review by tafaRU — APPROVED** [review]
Code review and test.

File: sale_order_type/models/sale_order.py
Comment (SimoRubi) [reply sale_order_type/models/sale_order.py:39]: @eLBati fields in `onchange` cannot be dotted

File: sale_order_type/readme/USAGE.rst
Comment (SimoRubi) [reply sale_order_type/readme/USAGE.rst:2]: Done in https://github.com/OCA/sale-workflow/pull/716/commits/148e7b60af81508b48788a45021d44f6d8e15ca3

File: sale_order_type/views/sale_order_view.xml
Comment (SimoRubi) [reply sale_order_type/views/sale_order_view.xml:13]: Done, along with some refactoring, in https://github.com/OCA/sale-workflow/pull/716/commits/148e7b60af81508b48788a45021d44f6d8e15ca3

**Comment by SimoRubi** [comment]
@tafaRU and @eLBati thanks for your reviews, I did what you asked for in https://github.com/OCA/sale-workflow/pull/716/commits/148e7b60af81508b48788a45021d44f6d8e15ca3, could you please check?

File: sale_order_type/models/sale_order_type_rule.py
Comment (tafaRU) [inline sale_order_type/models/sale_order_type_rule.py]: s/if/iff

File: sale_order_type/models/sale_order_type_rule.py
Comment (tafaRU) [inline sale_order_type/models/sale_order_type_rule.py]: same as above

File: sale_order_type/models/sale_order_type_rule.py
Comment (tafaRU) [inline sale_order_type/models/sale_order_type_rule.py]: same as above

File: sale_order_type/models/sale_order_type_rule.py
Comment (tafaRU) [inline sale_order_type/models/sale_order_type_rule.py]: same as above

File: sale_order_type/__manifest__.py
Comment (tafaRU) [inline sale_order_type/__manifest__.py:22]: why did you remove `account` from `depends`?

**Review by tafaRU — CHANGES_REQUESTED** [review]

File: sale_order_type/__manifest__.py
Comment (SimoRubi) [reply sale_order_type/__manifest__.py:22]: @tafaRU because it is redundant: it is already included because `sale_stock` depends on `sale` that depends on `account`

File: sale_order_type/models/sale_order_type_rule.py
Comment (SimoRubi) [reply sale_order_type/models/sale_order_type_rule.py]: @tafaRU fixed in https://github.com/OCA/sale-workflow/pull/716/commits/401a71d856124291b13544f84ec2632c99762f26, same for below

**Comment by SimoRubi** [comment]
@eLBati I did what you asked for could you please check and update your review?

File: sale_order_type/README.rst
Comment (eLBati) [inline sale_order_type/README.rst:47]: USAGE.rst is not included here. Could you run oca-gen-addon-readme?

**Review by eLBati — CHANGES_REQUESTED** [review]
Created an order type
image
where ice rule is
image
created an invoice
image

Payment term should be `30% Advance End of Following Month`. It is not correctly computed by onchange nor by button

**Comment by SimoRubi** [comment]
@eLBati I fixed the bug you found out in https://github.com/OCA/sale-workflow/pull/716/commits/9ea90328bb6514bb1f54c7ef899bafc2e493e11a, could you please check now?
@tafaRU I did what you asked for, could you please update your review?
Thanks

File: sale_order_type/models/account_invoice.py
Comment (tafaRU) [inline sale_order_type/models/account_invoice.py:27]: Good catch!

Comment (SimoRubi) [reply sale_order_type/models/account_invoice.py:27]: That is thanks to https://github.com/OCA/sale-workflow/pull/716#pullrequestreview-175395682

**Review by tafaRU — APPROVED** [review]
This time I only did the code review.

**Comment by eLBati** [comment]
@SimoRubi ok, thanks.
Please just use user friendly labels for  `product ids` and `product category ids` 

image

**Comment by SimoRubi** [comment]
@eLBati done in https://github.com/OCA/sale-workflow/pull/716/commits/bc79830e81636ab173522360f9531e383c31b286

File: sale_order_type/README.rst
Comment (SimoRubi) [reply sale_order_type/README.rst:47]: Done in https://github.com/OCA/sale-workflow/pull/716/commits/bc79830e81636ab173522360f9531e383c31b286

**Review by eLBati — APPROVED** [review]
This should be ported to 11.0 afterwards

**Comment by SimoRubi** [comment]
Fixed commit history

**Comment by SimoRubi** [comment]
@eLBati I'll try to cherry-pick this commit in v11 after it's merged

**Comment by SimoRubi** [comment]
@OCA/crm-sales-marketing-maintainers can this be merged? Thanks

**Comment by pedrobaeza** [comment]
Instead of using the sequence, isn't this interesting to be auto-sorted the same as now pricelist items are auto-arranged?

**Comment by SimoRubi** [comment]
@pedrobaeza I don't know what auto-arrangement you are suggesting, could you provide a link where it is performed? It sounds interesting, thanks

**Comment by pedrobaeza** [comment]
I mean this:

https://github.com/odoo/odoo/blob/aed0a828e4c6c1945c09f9701b2e5e5992954924/addons/product/models/product_pricelist.py#L365

in combination with a proper sorted selection field:

https://github.com/odoo/odoo/blob/aed0a828e4c6c1945c09f9701b2e5e5992954924/addons/product/models/product_pricelist.py#L384

This selection field can be computed stored and got from the input in the rest of the fields.

**Comment by SimoRubi** [comment]
Thanks @pedrobaeza for the suggestion, that is another good thing to know about Odoo.
In this case, I prefer to give the users more power (it comes with more responsibility) and let them choose the order manually.
I added your suggestion in the roadmap of the module as a possible improvement.

**Pull request merged**
